Mn 2018 01 fix richtext field export errors

MagdaN opened the pull request:
It seems to me that the RichTextField saves the data in ascii/html and not in utf8, therefore text from those fields is not exported correctly. Using lxml fixed the problem for me with excel. 

But I am not sure if there could be a better solution for this as lxml is quite huge and by adding it here we would add to all our projects.

mkind commented:
Is there a connection to issue #213?

MagdaN commented:
@mkind not really, I would say, as this code is only used for exporting the textfield data to excel, and removes all html related characters from text, not just the ones which are not allowed

mkind commented on adhocracy4/exports/views.py:
```diff
@@ -84,14 +88,27 @@ def get_field_data(self, item, name):
-        return strip_tags(item.description).strip()
+    def get_rich_text_field_data(self, text):
+        html = lxml.html.fromstring(text)
+        return html.text_content()
```
is the result the same as calling `clean_html_all` from [transforms.py](https://github.com/liqd/adhocracy4/blob/master/adhocracy4/transforms.py#L6)?

kleingeist commented:
> Is there a connection to issue #213?
@mkind no issue #213 is to get rid of the need to call the html cleaning manually in the `save` method. this is related to the export.

but i don't quit eget the problem or the proposed solution. if it is an encoding error we could just encode via python stdlib functions.

if it is things like `&gt;` instead of `>` i don't get how lxml fixes the problem.

kleingeist commented on adhocracy4/exports/views.py:
```diff
@@ -84,14 +88,27 @@ def get_field_data(self, item, name):
+
+        fieldtype = self._get_field_type(item, name)
+
+        if fieldtype and fieldtype == 'RichTextField':
```
i don't like the approach to make the `get_field_data` even more complicated. i think it would be better to put logic like the html cleaning to the concrete get_X_data method instead.
maybe we could also remove the call to the `get_X_display` and the logic to handle dicts, too (especially as i realized, that that the call `return item['name']` should be `return item[name]` instead. which we never realized as the the export is never called on lists of dicts)

MagdaN replied:
I was thinking about this, but I was afraid, that we would then miss this at some point, because we forget or don't know that a field is a RichTextFiled

kleingeist replied:
but if it is the html encodings then the problem occurs for comments, too

MagdaN replied:
No, because it is not a RichTextField

kleingeist replied:
then, as guesse above, i don't get the problem that is fixed with this pr
if i enter `greater > test` as a comment it gets stored as `greater &gt; test` to the database and exportet as `greater &gt; test` in the comments

kleingeist commented on adhocracy4/exports/views.py:
```diff
@@ -84,14 +88,27 @@ def get_field_data(self, item, name):
 
-    def get_description_data(self, item):
-        return strip_tags(item.description).strip()
+    def get_rich_text_field_data(self, text):
```
if we keep this general method i would prefer to rename it, as the naming scheme was `get_X_data` where `X` is standing for a (virtual) field which is represented as a column in the export.

MagdaN replied:
Sure, what would you suggest? We could extend this for more fields

MagdaN commented:
I added this based on: 
https://stackoverflow.com/questions/37486/filter-out-html-tags-and-resolve-entities-in-python

MagdaN commented:
@kleingeist works nicely with html from python, thanks! I removed get_field_data and changed moved the functationality back to get_description_data but was wondering if get_description_data should actually be in A4

kleingeist commented on adhocracy4/exports/mixins.py:
```diff
@@ -87,14 +85,14 @@ def _flat_comments(self, item):
                     date=reply.created.isoformat(),
                     username=reply.creator.username,
-                    text=strip_tags(reply.comment).strip()
+                    text=self.strip_and_unescape_html(reply.comment).strip()
```
the last `strip()` is not needed

kleingeist commented on adhocracy4/exports/views.py:
```diff
@@ -67,7 +67,12 @@ def get_base_filename(self):
 
+    def strip_and_unescape_html(self, text):
+        text = strip_tags(text).strip()
+        return html.unescape(text)
```
sorry for all those comments:
it would make sense to call `html.unescape` prior to`strip()` as it might replace `&nbsp;` with actual whitespaces which then may be removed by strip

MagdaN commented:
@kleingeist changed it
